reacting_rl_envs/hit_target.py

```diff
@@ -0,0 +1,272 @@
+
+    The environment is partially observable - the controlled object does not know the layout of the obstacles.
+
+    The action specifies angular velocity of the controlled object. The actual speed is not controllable
+
+    The observation contains:
+    object location x
+    object location y
+    object direction in range [-pi, pi)
+    target location x
+    target location y
+    closeness (1/(1+dist)) in range [0-1] of the nearest obstacle in each view:
+        0 - no obstacle in range for the given view
+        1 - the obstacle is next to the controlled object
+    """
+
+    ObsType = Union[np.ndarray, dict]
+
+    def __init__(
+            self, view_range=0.2, num_views=8,
+            object_size=0.025, num_obstacles=20, obstacle_size=0.05,
+            speed=0.005, max_angular_velocity=0.2, min_obstacle_dist=0.1,
+            target_dist = 0.9,
+            flat_obs=False) -> None:
+        self.num_views = num_views
+        self.view_range = view_range
+        self.object_size=object_size
+        self.num_obstacles=num_obstacles
+        self.obstacle_size=obstacle_size
+        self.speed = speed
+        self.max_angular_velocity = max_angular_velocity
+        self.min_obstacle_dist = min_obstacle_dist
+        self.target_dist = target_dist
+        self.flat_obs = flat_obs
+
+        self.action_space = Box(-1, 1)
+        low_obs = np.array([-np.inf, -np.inf, -np.pi] + [0] * num_views)
+        low_goal = np.array([-np.inf, -np.inf])
+        high_obs = np.array([np.inf, np.inf, np.pi] + [1] * num_views)
+        high_goal = np.array([np.inf, np.inf])
+        if flat_obs:
+            self.observation_space = Box(
+                np.concatenate([low_obs, low_goal]),
+                np.concatenate([high_obs, high_goal])
+            )
+        else:
+            self.observation_space = Dict({
+                'observation': Box(low_obs, high_obs),
+                'desired_goal': Box(low_goal, high_goal),
+                'achieved_goal': Box(low_goal, high_goal),
+            })
+
+
+        self.world = b2World(gravity=(0, 0))
+
+        self.object = self.world.CreateDynamicBody(
+            position=(0, 0),
+            fixtures=b2FixtureDef(
+                shape=b2CircleShape(pos=(0, 0), radius=object_size),
+                restitution=0,
+                density=0,
+            )
+        )
+        self.target = np.array([1, 0])
+        self.obstacles = [self.world.CreateStaticBody(
+            position=(1, 1),
+            fixtures=b2FixtureDef(shape=b2CircleShape(pos=(0, 0), radius=obstacle_size))
+        ) for _ in range(num_obstacles)]
+
+        self.window = None
+
+    def step(self, action: float) -> Tuple[ObsType, float, bool, bool, dict]:
+        angle_diff = float(action * self.max_angular_velocity)
+        angle = get_angle(self.object.linearVelocity)
+        self.object.linearVelocity = np.cos(angle + angle_diff) * self.speed, np.sin(angle + angle_diff) * self.speed
```

Comment:
I cannot find the logic of reducing the rocket speed because of collision?

Reply:
There is no such logic; instead the Box2D collisions are used.